src/NuGet.Services.AzureSearch/DocumentUtilities.cs

```diff
@@ -0,0 +1,216 @@
+            document.Summary = leaf.Summary;
+            document.Tags = leaf.Tags == null ? null : leaf.Tags.ToArray();
+            document.Title = leaf.Title;
+        }
+
+        public static string GetSearchDocumentKey(string packageId, SearchFilters searchFilters)
+        {
+            var lowerId = packageId.ToLowerInvariant();
+            var encodedId = EncodeKey(lowerId);
+            return $"{encodedId}-{searchFilters}";
+        }
+
+        public static string GetHijackDocumentKey(string packageId, string normalizedVersion)
+        {
+            var lowerId = packageId.ToLowerInvariant();
+            var lowerVersion = normalizedVersion.ToLowerInvariant();
+            return EncodeKey($"{lowerId}/{lowerVersion}");
+        }
+
+        private static string EncodeKey(string rawKey)
+        {
+            // First, encode the raw value for uniqueness.
+            var bytes = Encoding.UTF8.GetBytes(rawKey);
+            var unique = HttpServerUtility.UrlTokenEncode(bytes);
+
+            // Then, prepend a string as close to the raw key as possible, for readibility.
+            var readable = ReplaceUnsafeKeyCharacters(rawKey).TrimStart('_');
+
+            return readable.Length > 0 ? $"{readable}-{unique}" : unique;
+        }
+
+        private static string ReplaceUnsafeKeyCharacters(string input)
+        {
+            return Regex.Replace(
+                input,
+                "[^A-Za-z0-9-_]", // Remove equal sign as well, since it's ugly.
+                "_",
+                RegexOptions.None,
+                TimeSpan.FromSeconds(30));
+        }
+
+        private static DateTimeOffset? AssumeUtc(DateTime? dateTime)
+        {
+            if (!dateTime.HasValue)
+            {
+                return null;
+            }
+
+            return new DateTimeOffset(
+                DateTime.SpecifyKind(dateTime.Value, DateTimeKind.Utc),
+                TimeSpan.Zero);
+        }
+
+        private static string GetFlattenedDependencies(PackageDetailsCatalogLeaf leaf)
+        {
+            if (leaf.DependencyGroups == null)
+            {
+                return null;
+            }
+
+            var builder = new StringBuilder();
+            foreach (var dependencyGroup in leaf.DependencyGroups)
+            {
+                var targetFramework = dependencyGroup.ParseTargetFramework();
+
+                if (dependencyGroup.Dependencies != null && dependencyGroup.Dependencies.Any())
+                {
+                    foreach (var packageDependency in dependencyGroup.Dependencies)
+                    {
+                        AddFlattennedPackageDependency(targetFramework, packageDependency, builder);
+                    }
+                }
+                else
+                {
+                    if (builder.Length > 0)
+                    {
+                        builder.Append("|");
+                    }
+
+                    builder.Append(":");
```

Comment:
It would be good to add documentation --- internal is fine --- about this custom serialization format.